File: Lab/Lab1-Python/blend_float.py
"""
4190.308 Computer Architecture                                                          Spring 2023

Image blending (float)

This module implements a function that blends two images together (floating point version)

@author:
    Your Name <your@email.com>

@changes:
    2023/MM/DD Your Name Comment

"""



def blend(img1, img2, height, width, channels, overlay, alpha):
    """
    Alpha-blends two images of size heightxwidth. The image data must contain an alpha
    channel, i.e., 'channels' must be four

    Args:
        img1:         image 1 data (multi-level list)
        img2:         image 2 data (multi-level list)
        height:       image height
        width:        image width
        channels:     number of channels (must be BGRA)
        overlay:      if 1, overlay the second image over the first
                      if 0, merge the two images
        alpha:        alpha blending factor (0.0-1.0)

    Returns:
        A tuple containing the following elements:
        - blended:    blended image data (multi-level list)
        - bheight:    blended image height (=height)
        - bwidth:     blended image width (=width)
        - bchannels:  blended image channels (=channels)

    """

    if channels != 4:
        raise ValueError('Invalid number of channels')


    # TODO
    # Your work goes here

    # For now, we simply copy the input parameters into the output parameters.
    # Fix/adjust once you have implemented your solution.
    bheight   = height
    bwidth    = width
    bchannels = channels
    blended   = [[[0 for c in range(bchannels)] for w in range(bwidth)] for h in range(bheight)]

    for h in range(height):
        for w in range(width):
            for c in range(channels):
                if overlay == 1:
                    if c != 3:
                        overlay_ratio = (img2[h][w][3] / 255.0) * alpha
                            
                        multi1 = (img1[h][w][c] / 255.0) * (1 - overlay_ratio)
                        multi2 = (img2[h][w][c] / 255.0) * overlay_ratio
                        blended[h][w][c] = int((multi1 + multi2) * 255.0)
                    else:
                        blended[h][w][c] = img1[h][w][c]
                else:
                    if c != 3:
                        multi1 = (img1[h][w][c] / 255.0) * (img1[h][w][3] / 255.0) * (1 - alpha)
                        multi2 = (img2[h][w][c] / 255.0) * (img2[h][w][3] / 255.0) * alpha
                        blended[h][w][c] = int((multi1 + multi2) * 255.0)
                    else:
                        multi1 = (img1[h][w][c] / 255.0) * (1 - alpha)
                        multi2 = (img2[h][w][c] / 255.0) * alpha
                        blended[h][w][c] = int((multi1 + multi2) * 255.0)

    return blended, bheight, bwidth, bchannels

File: Lab/Lab2-C/blend_float.c
//-------------------------------------------------------------------------------------------------
// 4190.308 Computer Architecture                                                       Spring 2023
//
/// @file
/// @brief Image blending (float)
///        This module implements a function that blends two images together filter (floating-point
///        version). Note that CPython uses the 'double' data type for floating point numbers, 
///        so one may observe small differences in the output if the 'float' data type is used.
///
/// @author Bae Munsung <hamasoon@snu.ac.kr>
///
/// @section changelog Change Log
/// 2023/MM/DD Your Name Comment
///
//-------------------------------------------------------------------------------------------------

#include <stdlib.h>
#include "blend.h"


struct Image blend_float(struct Image img1, struct Image img2, int overlay, double alpha)
{
    if (img1.channels != 4) abort();


    // Initialize blended image
    struct Image blended = {
        .height   = img1.height,
        .width    = img1.width,
        .channels = img1.channels
    };
    blended.data = malloc(blended.height*blended.width*blended.channels*sizeof(uint8));
    if (blended.data == NULL) abort();


    for (size_t i = 0; i < blended.height; i++) {
        for (size_t j = 0; j < blended.width; j++) {
            for (size_t k = 0; k < blended.channels; k++) {
                if(overlay) {
                    if(k != 3) {
                        double overlay_ratio = img2.data[3 + blended.channels*(j + i*blended.width)] * alpha / 255.0;
                        double temp = img1.data[k + blended.channels*(j + i*blended.width)] * (1 - overlay_ratio);
                        temp += img2.data[k + blended.channels*(j + i*blended.width)] * overlay_ratio;
                            
                        blended.data[k + blended.channels*(j + i*blended.width)] = (uint8)(temp);
                    }
                    else {
                        blended.data[k + blended.channels*(j + i*blended.width)] = img1.data[k + blended.channels*(j + i*blended.width)];
                    }
                }
                else {
                    if(k != 3) {
                        double temp = (img1.data[k + blended.channels*(j + i*blended.width)] * img1.data[3 + blended.channels*(j + i*blended.width)] * (1 - alpha));
                        temp += (img2.data[k + blended.channels*(j + i*blended.width)] * img2.data[3 + blended.channels*(j + i*blended.width)] * alpha);
                        blended.data[k + blended.channels*(j + i*blended.width)] = (uint8)(temp / 255.0);
                    }
                    else {
                        double temp = (img1.data[3 + blended.channels*(j + i*blended.width)] * (1 - alpha));
                        temp += (img2.data[3 + blended.channels*(j + i*blended.width)] * alpha);
                        blended.data[k + blended.channels*(j + i*blended.width)] = (uint8)(temp);
                    }
                }
            }
        }
    }


    return blended;
}
